apps/src/dance/ai/AiBlockPreview.tsx

```diff
@@ -18,7 +18,23 @@ const AiBlockPreview: React.FunctionComponent<AiBlockPreviewProps> = ({
 }) => {
   const blockPreviewContainerRef = useRef<HTMLSpanElement>(null);
   const refTimer = useRef<number | null>(null);
+  const workspaceRef = useRef<Workspace | null>(null);
 
+  // Create the workspace once the container has been rendered.
```

Comment:
@bencodeorg you had the [right idea](https://github.com/code-dot-org/code-dot-org/pull/54602#discussion_r1378004657) here :) turns out that if a workspace has no parent workspace, calling `dispose()` will not only remove the workspace from the DOM, but will also remove its parent element, because Blockly assumes it's the top level workspace. Not sure why exactly that's the case, but it became an issue here since calling `dispose()` would remove the `<span>` that the workspace was rendered into, so subsequent workspaces were being created with empty elements.

I updated this along the lines of what you suggested - creating a single workspace once the container has mounted, and switched to using `workspace.clear()` as the cleanup effect, which will just remove all blocks. We still call `workspace.dispose()` but only when the entire component is unmounting.

Reply:
Oh nice, good catch!

Reply:
Yeah, looks like that happens in WorkspaceSvg [here](https://github.com/google/blockly/blob/9d117674ce64e88db35edd4426334fa4ae23d832/core/workspace_svg.ts#L887). Seems to be at least intentional but I'm curious what the rationale is